File: 020_valid_parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        int len = s.length();
        if(len == 0){
            return true;
        }
        if(s[0] == '}' || s[0] == ']' || s[0] == ')'){
            return false;
        }
        map<char, char> m = {{']','['}, {'}','{'}, {')','('}};
        vector<int> stack;
        stack.push_back(s[0]);
        char c;
        for(int i = 1; i < len; i++){
            if(s[i] == '}' || s[i] == ']' || s[i] == ')'){
                if(stack.size() == 0) return false;
                else{
                    c = stack[stack.size() - 1];
                    stack.pop_back();
                    if(c != m[s[i]]) return false;
                }
            }
            else stack.push_back(s[i]);
        }
        if(stack.size() == 0) return true;
        else return false;
    }
};

File: 020_valid_parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if s == "":
            return True
        if s[0] == '}' or s[0] == ']' or s[0] == ')':
            return False
        d = {')': '(', ']': '[', '}': '{'}
        stack = collections.deque()
        stack.append(s[0])
        for i in range(1, len(s)):
            if s[i] == ')' or s[i] == '}' or s[i] == ']':
                if len(stack) == 0:
                    return False
                else:
                    c = stack.pop()
                    if d[s[i]] != c:
                        return False
            else:
                stack.append(s[i])
        if len(stack) == 0:
            return True
        else:
            return False

            
